In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('/opt/ml/project2/level2_klue_nlp-level2-nlp-01/dataset/train/train.csv')

In [3]:
df

id                                           sentence  \
0          0  〈Something〉는 조지 해리슨이 쓰고 비틀즈가 1969년 앨범 《Abbey R...   
1          1  호남이 기반인 바른미래당·대안신당·민주평화당이 우여곡절 끝에 합당해 민생당(가칭)으...   
2          2  K리그2에서 성적 1위를 달리고 있는 광주FC는 지난 26일 한국프로축구연맹으로부터...   
3          3  균일가 생활용품점 (주)아성다이소(대표 박정부)는 코로나19 바이러스로 어려움을 겪...   
4          4  1967년 프로 야구 드래프트 1순위로 요미우리 자이언츠에게 입단하면서 등번호는 8...   
...      ...                                                ...   
32465  32465  한국당은 7일 오전 9시부터 오후 5시까지 진행된 원내대표 및 정책위의장 후보자 등...   
32466  32466  법포는 다시 최시형, 서병학, 손병희 직계인 북접과 다시 서장옥, 전봉준, 김개남을...   
32467  32467  완도군(군수 신우철)이 국토교통부에서 실시한 '2019 교통문화지수 실태조사'에서 ...   
32468  32468  중앙일보, JTBC 회장을 지낸 이후 중앙홀딩스 회장, 재단법인 한반도평화만들기 이...   
32469  32469  화순군(군수 구충곤)은 17일 동면의 이장 20여 명이 코로나 19 예방을 위해 버...   

                                          subject_entity  \
0      {'word': '비틀즈', 'start_idx': 24, 'end_idx': 26...   
1      {'word': '민주평화당', 'start_idx': 19, 'end_idx': ...   
2      {'word': '광주FC', 'start_idx': 21, 'end_idx': 2...   
3      {'word': '아성다이소', 'start_idx': 13, 'end_idx': ...   
4      {'word': '요미우리 자이언츠', 'start_idx': 22, 'end_id...   
...                                                  ...   
32465  {'word': '유기준', 'start_idx': 93, 'end_idx': 95...   
32466  {'word': '최시형', 'start_idx': 7, 'end_idx': 9, ...   
32467  {'word': '완도군', 'start_idx': 0, 'end_idx': 2, ...   
32468  {'word': 'JTBC', 'start_idx': 6, 'end_idx': 9,...   
32469  {'word': '화순군', 'start_idx': 0, 'end_idx': 2, ...   

                                           object_entity  \
0      {'word': '조지 해리슨', 'start_idx': 13, 'end_idx':...   
1      {'word': '대안신당', 'start_idx': 14, 'end_idx': 1...   
2      {'word': '한국프로축구연맹', 'start_idx': 34, 'end_idx...   
3      {'word': '박정부', 'start_idx': 22, 'end_idx': 24...   
4      {'word': '1967', 'start_idx': 0, 'end_idx': 3,...   
...                                                  ...   
32465  {'word': '부산 서구·동구', 'start_idx': 100, 'end_id...   
32466  {'word': '손병희', 'start_idx': 17, 'end_idx': 19...   
32467  {'word': '신우철', 'start_idx': 7, 'end_idx': 9, ...   
32468  {'word': '중앙홀딩스', 'start_idx': 21, 'end_idx': ...   
32469  {'word': '구충곤', 'start_idx': 7, 'end_idx': 9, ...   

                           label     source  
0                    no_relation  wikipedia  
1                    no_relation   wikitree  
2                  org:member_of   wikitree  
3      org:top_members/employees   wikitree  
4                    no_relation  wikipedia  
...                          ...        ...  
32465            per:employee_of   wikitree  
32466             per:colleagues  wikipedia  
32467  org:top_members/employees   wikitree  
32468                no_relation  wikipedia  
32469  org:top_members/employees   wikitree  

[32470 rows x 6 columns]

# 'sentence', 'subject_entity','object_entity','label' 중복 데이터 확인 및 제거

In [4]:
duplicated_df=df[df.duplicated(['sentence', 'subject_entity','object_entity','label'], keep=False)]

In [5]:
duplicated_df.sort_values(by=['sentence'])

id                                           sentence  \
22090  22090  1959-60 시즌 이후의 알파 에트니키 / 수페르리가 엘라다, 그리스 컵, 그리고...   
9618    9618  1959-60 시즌 이후의 알파 에트니키 / 수페르리가 엘라다, 그리스 컵, 그리고...   
25368  25368  2001년 4월 민주국민당 김윤환 의원과 민주당 의원 사이의 정책 연합과 관련한 통...   
2500    2500  2001년 4월 민주국민당 김윤환 의원과 민주당 의원 사이의 정책 연합과 관련한 통...   
3547    3547  2010년 10월 15일 전라북도청을 방문한 맹형규 행정안전부 장관은 도내 지자체간...   
...      ...                                                ...   
27325  27325    펀치넬로는 10cm의 권정열이 피처링으로 참여한 '미안해서 미안해'로 무대에 올랐다.   
14658  14658                              함연지는 함영준 오뚜기 회장의 장녀다.   
6936    6936                              함연지는 함영준 오뚜기 회장의 장녀다.   
20898  20898  화려한 재기를 뒤로하고 고(故) 안재환에게 돈을 빌려줬다는 루머로 마음고생을 하던 ...   
4248    4248  화려한 재기를 뒤로하고 고(故) 안재환에게 돈을 빌려줬다는 루머로 마음고생을 하던 ...   

                                          subject_entity  \
22090  {'word': '수페르리가 엘라다', 'start_idx': 25, 'end_id...   
9618   {'word': '수페르리가 엘라다', 'start_idx': 25, 'end_id...   
25368  {'word': '민주국민당', 'start_idx': 9, 'end_idx': 1...   
2500   {'word': '민주국민당', 'start_idx': 9, 'end_idx': 1...   
3547   {'word': '전라북도청', 'start_idx': 14, 'end_idx': ...   
...                                                  ...   
27325  {'word': '권정열', 'start_idx': 12, 'end_idx': 14...   
14658  {'word': '오뚜기', 'start_idx': 9, 'end_idx': 11,...   
6936   {'word': '오뚜기', 'start_idx': 9, 'end_idx': 11,...   
20898  {'word': '최진실', 'start_idx': 46, 'end_idx': 48...   
4248   {'word': '최진실', 'start_idx': 46, 'end_idx': 48...   

                                           object_entity  \
22090  {'word': '그리스', 'start_idx': 36, 'end_idx': 38...   
9618   {'word': '그리스', 'start_idx': 36, 'end_idx': 38...   
25368  {'word': '민주당', 'start_idx': 23, 'end_idx': 25...   
2500   {'word': '민주당', 'start_idx': 23, 'end_idx': 25...   
3547   {'word': '행정안전부', 'start_idx': 29, 'end_idx': ...   
...                                                  ...   
27325  {'word': '10cm', 'start_idx': 6, 'end_idx': 9,...   
14658  {'word': '함영준', 'start_idx': 5, 'end_idx': 7, ...   
6936   {'word': '함영준', 'start_idx': 5, 'end_idx': 7, ...   
20898  {'word': '2008년 10월 2일', 'start_idx': 51, 'end...   
4248   {'word': '2008년 10월 2일', 'start_idx': 51, 'end...   

                           label     source  
22090                no_relation  wikipedia  
9618                 no_relation  wikipedia  
25368                no_relation  wikipedia  
2500                 no_relation  wikipedia  
3547                 no_relation  wikipedia  
...                          ...        ...  
27325            per:employee_of   wikitree  
14658  org:top_members/employees   wikitree  
6936   org:top_members/employees   wikitree  
20898          per:date_of_death   wikitree  
4248           per:date_of_death   wikitree  

[84 rows x 6 columns]

In [6]:
new_df = df.drop_duplicates(['sentence', 'subject_entity','object_entity','label'])

In [7]:
new_df

id                                           sentence  \
0          0  〈Something〉는 조지 해리슨이 쓰고 비틀즈가 1969년 앨범 《Abbey R...   
1          1  호남이 기반인 바른미래당·대안신당·민주평화당이 우여곡절 끝에 합당해 민생당(가칭)으...   
2          2  K리그2에서 성적 1위를 달리고 있는 광주FC는 지난 26일 한국프로축구연맹으로부터...   
3          3  균일가 생활용품점 (주)아성다이소(대표 박정부)는 코로나19 바이러스로 어려움을 겪...   
4          4  1967년 프로 야구 드래프트 1순위로 요미우리 자이언츠에게 입단하면서 등번호는 8...   
...      ...                                                ...   
32465  32465  한국당은 7일 오전 9시부터 오후 5시까지 진행된 원내대표 및 정책위의장 후보자 등...   
32466  32466  법포는 다시 최시형, 서병학, 손병희 직계인 북접과 다시 서장옥, 전봉준, 김개남을...   
32467  32467  완도군(군수 신우철)이 국토교통부에서 실시한 '2019 교통문화지수 실태조사'에서 ...   
32468  32468  중앙일보, JTBC 회장을 지낸 이후 중앙홀딩스 회장, 재단법인 한반도평화만들기 이...   
32469  32469  화순군(군수 구충곤)은 17일 동면의 이장 20여 명이 코로나 19 예방을 위해 버...   

                                          subject_entity  \
0      {'word': '비틀즈', 'start_idx': 24, 'end_idx': 26...   
1      {'word': '민주평화당', 'start_idx': 19, 'end_idx': ...   
2      {'word': '광주FC', 'start_idx': 21, 'end_idx': 2...   
3      {'word': '아성다이소', 'start_idx': 13, 'end_idx': ...   
4      {'word': '요미우리 자이언츠', 'start_idx': 22, 'end_id...   
...                                                  ...   
32465  {'word': '유기준', 'start_idx': 93, 'end_idx': 95...   
32466  {'word': '최시형', 'start_idx': 7, 'end_idx': 9, ...   
32467  {'word': '완도군', 'start_idx': 0, 'end_idx': 2, ...   
32468  {'word': 'JTBC', 'start_idx': 6, 'end_idx': 9,...   
32469  {'word': '화순군', 'start_idx': 0, 'end_idx': 2, ...   

                                           object_entity  \
0      {'word': '조지 해리슨', 'start_idx': 13, 'end_idx':...   
1      {'word': '대안신당', 'start_idx': 14, 'end_idx': 1...   
2      {'word': '한국프로축구연맹', 'start_idx': 34, 'end_idx...   
3      {'word': '박정부', 'start_idx': 22, 'end_idx': 24...   
4      {'word': '1967', 'start_idx': 0, 'end_idx': 3,...   
...                                                  ...   
32465  {'word': '부산 서구·동구', 'start_idx': 100, 'end_id...   
32466  {'word': '손병희', 'start_idx': 17, 'end_idx': 19...   
32467  {'word': '신우철', 'start_idx': 7, 'end_idx': 9, ...   
32468  {'word': '중앙홀딩스', 'start_idx': 21, 'end_idx': ...   
32469  {'word': '구충곤', 'start_idx': 7, 'end_idx': 9, ...   

                           label     source  
0                    no_relation  wikipedia  
1                    no_relation   wikitree  
2                  org:member_of   wikitree  
3      org:top_members/employees   wikitree  
4                    no_relation  wikipedia  
...                          ...        ...  
32465            per:employee_of   wikitree  
32466             per:colleagues  wikipedia  
32467  org:top_members/employees   wikitree  
32468                no_relation  wikipedia  
32469  org:top_members/employees   wikitree  

[32428 rows x 6 columns]

# 'sentence', 'subject_entity','object_entity' 중복 데이터 확인 및 label 정리

In [8]:
duplicated_df=new_df[new_df.duplicated(['sentence', 'subject_entity','object_entity'], keep=False)]

In [9]:
duplicated_df.sort_values(by=['sentence','label'])

id                                           sentence  \
6749    6749  대한항공은 5일 조양호 회장의 3자녀가 보유한 싸이버스카이 주식 9만9900주 전량...   
12829  12829  대한항공은 5일 조양호 회장의 3자녀가 보유한 싸이버스카이 주식 9만9900주 전량...   
8364    8364  배우 김병철 씨가 연기하는 정복동은 천리마마트를 망하게 하기 위해 여러 계획을 세우...   
32299  32299  배우 김병철 씨가 연기하는 정복동은 천리마마트를 망하게 하기 위해 여러 계획을 세우...   
22258  22258  영화 '버즈 오브 프레이'는 배트맨이 없는 고담시에서 할리퀸, 헌트리스, 블랙 카나...   
11511  11511  영화 '버즈 오브 프레이'는 배트맨이 없는 고담시에서 할리퀸, 헌트리스, 블랙 카나...   
277      277  이날 프로그램 공개에서는 전북영산작법보존회와 김명신‧정상희의 사제동행 판소리, 관악...   
3296    3296  이날 프로그램 공개에서는 전북영산작법보존회와 김명신‧정상희의 사제동행 판소리, 관악...   
4212    4212  한편 전라남도는 최근 확진자가 발생한 순천시와 여수시에 마스크를 각각 2만장씩 총 ...   
25094  25094  한편 전라남도는 최근 확진자가 발생한 순천시와 여수시에 마스크를 각각 2만장씩 총 ...   

                                          subject_entity  \
6749   {'word': '대한항공', 'start_idx': 0, 'end_idx': 3,...   
12829  {'word': '대한항공', 'start_idx': 0, 'end_idx': 3,...   
8364   {'word': '정복동', 'start_idx': 15, 'end_idx': 17...   
32299  {'word': '정복동', 'start_idx': 15, 'end_idx': 17...   
22258  {'word': '배트맨', 'start_idx': 16, 'end_idx': 18...   
11511  {'word': '배트맨', 'start_idx': 16, 'end_idx': 18...   
277    {'word': '강태환', 'start_idx': 62, 'end_idx': 64...   
3296   {'word': '강태환', 'start_idx': 62, 'end_idx': 64...   
4212   {'word': '전라남도', 'start_idx': 3, 'end_idx': 6,...   
25094  {'word': '전라남도', 'start_idx': 3, 'end_idx': 6,...   

                                           object_entity  \
6749   {'word': '조양호', 'start_idx': 9, 'end_idx': 11,...   
12829  {'word': '조양호', 'start_idx': 9, 'end_idx': 11,...   
8364   {'word': '김병철', 'start_idx': 3, 'end_idx': 5, ...   
32299  {'word': '김병철', 'start_idx': 3, 'end_idx': 5, ...   
22258  {'word': '고담시', 'start_idx': 24, 'end_idx': 26...   
11511  {'word': '고담시', 'start_idx': 24, 'end_idx': 26...   
277    {'word': '색소폰', 'start_idx': 58, 'end_idx': 60...   
3296   {'word': '색소폰', 'start_idx': 58, 'end_idx': 60...   
4212   {'word': '여수시', 'start_idx': 26, 'end_idx': 28...   
25094  {'word': '여수시', 'start_idx': 26, 'end_idx': 28...   

                           label     source  
6749                 no_relation  wikipedia  
12829  org:top_members/employees  wikipedia  
8364                 no_relation   wikitree  
32299        per:alternate_names   wikitree  
22258                no_relation   wikitree  
11511     per:place_of_residence   wikitree  
277                  no_relation   wikitree  
3296                   per:title   wikitree  
4212                 org:members   wikitree  
25094  org:place_of_headquarters   wikitree

### no_relation 제거 &
### id 4212, 25094 데이터 4212 데이터(members) 선택
### org:members : Organizations which belong to the specified organization
### org:place of headquarters : The place which the headquarters of the specified organization are located in

In [10]:
new_df = new_df.drop_duplicates(['sentence', 'subject_entity','object_entity'],keep='last')

In [11]:
new_df

id                                           sentence  \
0          0  〈Something〉는 조지 해리슨이 쓰고 비틀즈가 1969년 앨범 《Abbey R...   
1          1  호남이 기반인 바른미래당·대안신당·민주평화당이 우여곡절 끝에 합당해 민생당(가칭)으...   
2          2  K리그2에서 성적 1위를 달리고 있는 광주FC는 지난 26일 한국프로축구연맹으로부터...   
3          3  균일가 생활용품점 (주)아성다이소(대표 박정부)는 코로나19 바이러스로 어려움을 겪...   
4          4  1967년 프로 야구 드래프트 1순위로 요미우리 자이언츠에게 입단하면서 등번호는 8...   
...      ...                                                ...   
32465  32465  한국당은 7일 오전 9시부터 오후 5시까지 진행된 원내대표 및 정책위의장 후보자 등...   
32466  32466  법포는 다시 최시형, 서병학, 손병희 직계인 북접과 다시 서장옥, 전봉준, 김개남을...   
32467  32467  완도군(군수 신우철)이 국토교통부에서 실시한 '2019 교통문화지수 실태조사'에서 ...   
32468  32468  중앙일보, JTBC 회장을 지낸 이후 중앙홀딩스 회장, 재단법인 한반도평화만들기 이...   
32469  32469  화순군(군수 구충곤)은 17일 동면의 이장 20여 명이 코로나 19 예방을 위해 버...   

                                          subject_entity  \
0      {'word': '비틀즈', 'start_idx': 24, 'end_idx': 26...   
1      {'word': '민주평화당', 'start_idx': 19, 'end_idx': ...   
2      {'word': '광주FC', 'start_idx': 21, 'end_idx': 2...   
3      {'word': '아성다이소', 'start_idx': 13, 'end_idx': ...   
4      {'word': '요미우리 자이언츠', 'start_idx': 22, 'end_id...   
...                                                  ...   
32465  {'word': '유기준', 'start_idx': 93, 'end_idx': 95...   
32466  {'word': '최시형', 'start_idx': 7, 'end_idx': 9, ...   
32467  {'word': '완도군', 'start_idx': 0, 'end_idx': 2, ...   
32468  {'word': 'JTBC', 'start_idx': 6, 'end_idx': 9,...   
32469  {'word': '화순군', 'start_idx': 0, 'end_idx': 2, ...   

                                           object_entity  \
0      {'word': '조지 해리슨', 'start_idx': 13, 'end_idx':...   
1      {'word': '대안신당', 'start_idx': 14, 'end_idx': 1...   
2      {'word': '한국프로축구연맹', 'start_idx': 34, 'end_idx...   
3      {'word': '박정부', 'start_idx': 22, 'end_idx': 24...   
4      {'word': '1967', 'start_idx': 0, 'end_idx': 3,...   
...                                                  ...   
32465  {'word': '부산 서구·동구', 'start_idx': 100, 'end_id...   
32466  {'word': '손병희', 'start_idx': 17, 'end_idx': 19...   
32467  {'word': '신우철', 'start_idx': 7, 'end_idx': 9, ...   
32468  {'word': '중앙홀딩스', 'start_idx': 21, 'end_idx': ...   
32469  {'word': '구충곤', 'start_idx': 7, 'end_idx': 9, ...   

                           label     source  
0                    no_relation  wikipedia  
1                    no_relation   wikitree  
2                  org:member_of   wikitree  
3      org:top_members/employees   wikitree  
4                    no_relation  wikipedia  
...                          ...        ...  
32465            per:employee_of   wikitree  
32466             per:colleagues  wikipedia  
32467  org:top_members/employees   wikitree  
32468                no_relation  wikipedia  
32469  org:top_members/employees   wikitree  

[32423 rows x 6 columns]

In [12]:
new_df = new_df.append(duplicated_df[duplicated_df['id']==4212])

<ipython-input-12-55c70ed68485>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  new_df = new_df.append(duplicated_df[duplicated_df['id']==4212])


In [13]:
new_df

id                                           sentence  \
0          0  〈Something〉는 조지 해리슨이 쓰고 비틀즈가 1969년 앨범 《Abbey R...   
1          1  호남이 기반인 바른미래당·대안신당·민주평화당이 우여곡절 끝에 합당해 민생당(가칭)으...   
2          2  K리그2에서 성적 1위를 달리고 있는 광주FC는 지난 26일 한국프로축구연맹으로부터...   
3          3  균일가 생활용품점 (주)아성다이소(대표 박정부)는 코로나19 바이러스로 어려움을 겪...   
4          4  1967년 프로 야구 드래프트 1순위로 요미우리 자이언츠에게 입단하면서 등번호는 8...   
...      ...                                                ...   
32466  32466  법포는 다시 최시형, 서병학, 손병희 직계인 북접과 다시 서장옥, 전봉준, 김개남을...   
32467  32467  완도군(군수 신우철)이 국토교통부에서 실시한 '2019 교통문화지수 실태조사'에서 ...   
32468  32468  중앙일보, JTBC 회장을 지낸 이후 중앙홀딩스 회장, 재단법인 한반도평화만들기 이...   
32469  32469  화순군(군수 구충곤)은 17일 동면의 이장 20여 명이 코로나 19 예방을 위해 버...   
4212    4212  한편 전라남도는 최근 확진자가 발생한 순천시와 여수시에 마스크를 각각 2만장씩 총 ...   

                                          subject_entity  \
0      {'word': '비틀즈', 'start_idx': 24, 'end_idx': 26...   
1      {'word': '민주평화당', 'start_idx': 19, 'end_idx': ...   
2      {'word': '광주FC', 'start_idx': 21, 'end_idx': 2...   
3      {'word': '아성다이소', 'start_idx': 13, 'end_idx': ...   
4      {'word': '요미우리 자이언츠', 'start_idx': 22, 'end_id...   
...                                                  ...   
32466  {'word': '최시형', 'start_idx': 7, 'end_idx': 9, ...   
32467  {'word': '완도군', 'start_idx': 0, 'end_idx': 2, ...   
32468  {'word': 'JTBC', 'start_idx': 6, 'end_idx': 9,...   
32469  {'word': '화순군', 'start_idx': 0, 'end_idx': 2, ...   
4212   {'word': '전라남도', 'start_idx': 3, 'end_idx': 6,...   

                                           object_entity  \
0      {'word': '조지 해리슨', 'start_idx': 13, 'end_idx':...   
1      {'word': '대안신당', 'start_idx': 14, 'end_idx': 1...   
2      {'word': '한국프로축구연맹', 'start_idx': 34, 'end_idx...   
3      {'word': '박정부', 'start_idx': 22, 'end_idx': 24...   
4      {'word': '1967', 'start_idx': 0, 'end_idx': 3,...   
...                                                  ...   
32466  {'word': '손병희', 'start_idx': 17, 'end_idx': 19...   
32467  {'word': '신우철', 'start_idx': 7, 'end_idx': 9, ...   
32468  {'word': '중앙홀딩스', 'start_idx': 21, 'end_idx': ...   
32469  {'word': '구충곤', 'start_idx': 7, 'end_idx': 9, ...   
4212   {'word': '여수시', 'start_idx': 26, 'end_idx': 28...   

                           label     source  
0                    no_relation  wikipedia  
1                    no_relation   wikitree  
2                  org:member_of   wikitree  
3      org:top_members/employees   wikitree  
4                    no_relation  wikipedia  
...                          ...        ...  
32466             per:colleagues  wikipedia  
32467  org:top_members/employees   wikitree  
32468                no_relation  wikipedia  
32469  org:top_members/employees   wikitree  
4212                 org:members   wikitree  

[32424 rows x 6 columns]

In [14]:
new_df.to_csv('/opt/ml/project2/level2_klue_nlp-level2-nlp-01/dataset/train/new_train.csv', index=False)